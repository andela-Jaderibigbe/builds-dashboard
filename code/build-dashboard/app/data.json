[
  {
    "_id": 1,
    "user": "",
    "build_name": "Tenrox-R1_1235",
    "state": "pending",
    "progress": {
      "metrics": 0,
      "build": 0,
      "unit_test": 0,
      "functional_test": 0
    },
    "stats": {
      "metrics": null,
      "build": null,
      "unit_test": null,
      "functional_test": null
    },
    "time_started": null
  },
  {
    "_id": 2,
    "user": "jack",
    "build_name": "432462",
    "state": "running",
    "progress": {
      "metrics": 100,
      "build": 20,
      "unit_test": 0,
      "functional_test": 0
    },
    "stats": {
      "metrics": {
        "test": 64,
        "maintainability": 53,
        "security": 64,
        "workmanship": 72
      },
      "build": null,
      "unit_test": null,
      "functional_test": null
    },
    "time_started": "2017-10-17T15:48:12.600Z"
  },
  {
    "_id": 3,
    "user": "samy",
    "build_name": "432461",
    "state": "rejected",
    "progress": {
      "metrics": 100,
      "build": 100,
      "unit_test": 100,
      "functional_test": 100
    },
    "stats": {
      "metrics": {
        "test": 64,
        "maintainability": -53,
        "security": 64,
        "workmanship": 72
      },
      "build": {
        "build_date": "2017-10-15T10:48:12.600Z"
      },
      "unit_test": {
        "coverage": 0.76,
        "test_passing": 0.73,
        "covered": 142,
        "missed": 10
      },
      "functional_test": {
        "coverage": 0.23,
        "test_passing": 0.68,
        "covered": 4321,
        "missed": 2145
      }
    },
    "time_started": "2017-10-13T13:48:12.600Z"
  },
  {
    "_id": 4,
    "user": "",
    "build_name": "Tenrox-R1_1234",
    "state": "complete",
    "progress": {
      "metrics": 100,
      "build": 100,
      "unit_test": 100,
      "functional_test": 100
    },
    "stats": {
      "metrics": {
        "test": 64,
        "maintainability": 53,
        "security": 64,
        "workmanship": 72
      },
      "build": {
        "build_date": "2017-10-12T14:48:12.600Z"
      },
      "unit_test": {
        "coverage": 0.76,
        "test_passing": 0.73,
        "covered": 142,
        "missed": 10
      },
      "functional_test": {
        "coverage": 0.23,
        "test_passing": 0.68,
        "covered": 4321,
        "missed": 2145
      }
    },
    "time_started": "2017-10-11T11:48:12.600Z"
  },
  {
    "_id": 5,
    "user": "samy",
    "build_name": "432460",
    "state": "rejected",
    "progress": {
      "metrics": 100,
      "build": 0,
      "unit_test": 0,
      "functional_test": 0
    },
    "stats": {
      "metrics": {
        "test": 64,
        "maintainability": -53,
        "security": 64,
        "workmanship": 72
      },
      "build": null,
      "unit_test": null,
      "functional_test": null
    },
    "time_started": "2017-07-11T15:48:12.600Z"
  },
  {
    "_id": 6,
    "user": "samy",
    "build_name": "432459",
    "state": "accepted",
    "progress": {
      "metrics": 100,
      "build": 100,
      "unit_test": 100,
      "functional_test": 100
    },
    "stats": {
      "metrics": {
        "test": 64,
        "maintainability": 53,
        "security": 64,
        "workmanship": 72
      },
      "build": {
        "build_date": "2017-09-13T18:48:12.600Z"
      },
      "unit_test": {
        "coverage": 0.76,
        "test_passing": 0.73,
        "covered": 142,
        "missed": 10
      },
      "functional_test": {
        "coverage": 0.23,
        "test_passing": 0.68,
        "covered": 4321,
        "missed": 2145
      }
    },
    "time_started": "2017-08-13T12:48:12.600Z"
  }
]
