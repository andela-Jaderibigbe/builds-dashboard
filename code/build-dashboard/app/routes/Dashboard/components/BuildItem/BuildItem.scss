@import 'theme/variables';

.tableRow {
  width: 98%;
  margin: 1%;
}

.changeListCol {
  width: 19%;
  text-align: left;
}

.ownerCol {
  text-align: center;
  width: 9%;
}

.timeCol {
  text-align: center;
  width: 14%;
}

.stateCol {
  text-align: center;
  width: 14%;
}

.statusCol {
  text-align: center;
  width: 11%;
}

.buildName {
  cursor: pointer;
}

.buildName:hover {
  text-decoration: underline;
}

.buildIcon {
  i {
    margin-right: 15px;
  }
}

.isRunning {
  color: $blue;
  border: 1px solid $blue;
  background-image: linear-gradient(
    to right,
    rgba($blue, 0.1) 0%,
    transparent 50%
  );
}

.isPending {
  color: $gray;
  border: 1px solid $gray;
  background-image: linear-gradient(
    to right,
    rgba($gray, 0.1) 0%,
    transparent 50%
  );
}

.isRejected {
  color: $red;
  border: 1px solid $red;
  background-image: linear-gradient(
    to right,
    rgba($red, 0.1) 0%,
    transparent 50%
  );
}

.isSuccess {
  color: $green;
  border: 1px solid $green;
  background-image: linear-gradient(
    to right,
    rgba($green, 0.1) 0%,
    transparent 50%
  );
}
