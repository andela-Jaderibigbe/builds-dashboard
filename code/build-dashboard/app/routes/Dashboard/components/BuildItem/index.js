import React from 'react';
import PropTypes from 'prop-types';
import ExpandableRow from 'HOCS/ExpandableRow';
import Cell from 'components/Cell';
import TableRow from 'components/TableRow';
import Progress from 'components/Progress';
import Stats from '../Stats';
import { parseDatetime } from 'utils';
import { stateToDisplay, hasDrop } from '../../selector';
import styles from './BuildItem.scss';

const RowWithDetails = ExpandableRow(TableRow, Stats);

const getIconClassName = state => {
  switch (state) {
    case 'running':
      return 'fa fa-hourglass-half';
    case 'pending':
      return 'fa fa-hourglass-start';
    default:
      return 'fa fa-hourglass-end';
  }
};

const getRowStatus = state => {
  switch (state) {
    case 'running':
      return styles.isRunning;
    case 'pending':
      return styles.isPending;
    case 'rejected':
      return styles.isRejected;
    default:
      return styles.isSuccess;
  }
};

const BuildItem = props => {
  const {
    _id,
    user,
    build_name,
    state,
    progress: { metrics, build, unit_test, functional_test },
    time_started,
    stats,
    isSelected,
    onSelectItem,
  } = props;

  return (
    <RowWithDetails
      isSelected={isSelected}
      stats={stats}
      progress={props.progress}
      className={`${styles.tableRow} ${getRowStatus(state)}`}
      state={state}
    >
      <Cell className={styles.changeListCol}>
        <span className={styles.buildIcon}>
          <i className={getIconClassName(state)} />
          <span className={styles.buildName} onClick={() => onSelectItem(_id)}>
            {build_name}
          </span>
        </span>
      </Cell>
      <Cell className={styles.ownerCol}>{user}</Cell>
      <Cell className={styles.timeCol}>{parseDatetime(time_started)}</Cell>
      <Cell className={styles.stateCol}>{stateToDisplay(state)}</Cell>
      <Cell className={styles.statusCol} isHidden={isSelected}>
        <Progress
          progressValue={metrics}
          progressState={state}
          hasError={hasDrop(stats['metrics'] || {})}
        />
      </Cell>
      <Cell className={styles.statusCol} isHidden={isSelected}>
        <Progress
          progressValue={build}
          progressState={state}
          hasError={hasDrop(stats['build'] || {})}
        />
      </Cell>
      <Cell className={styles.statusCol} isHidden={isSelected}>
        <Progress
          progressValue={unit_test}
          progressState={state}
          hasError={hasDrop(stats['unit_test'] || {})}
        />
      </Cell>
      <Cell className={styles.statusCol} isHidden={isSelected}>
        <Progress
          progressValue={functional_test}
          progressState={state}
          hasError={hasDrop(stats['functional_test'] || {})}
        />
      </Cell>
    </RowWithDetails>
  );
};

BuildItem.propTypes = {
  _id: PropTypes.number,
  user: PropTypes.string,
  build_name: PropTypes.string,
  state: PropTypes.string,
  progress: PropTypes.object,
  time_started: PropTypes.string,
  stats: PropTypes.object,
  isSelected: PropTypes.bool,
  onSelectItem: PropTypes.func,
};

export default BuildItem;
