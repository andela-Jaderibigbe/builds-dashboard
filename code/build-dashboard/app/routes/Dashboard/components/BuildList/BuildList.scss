@import 'theme/variables';

.tableBody {
  overflow-y: scroll;
  display: block;
  flex: 1;
}

.tableRow {
  width: 98%;
  margin: 1%;
  cursor: pointer;
}

.changeListCol {
  width: 19%;
  text-align: left;
}

.ownerCol {
  text-align: center;
  width: 9%;
}

.timeCol {
  text-align: center;
  width: 14%;
}

.stateCol {
  text-align: center;
  width: 14%;
}

.statusCol {
  text-align: center;
  width: 11%;
}

.buildIcon {
  i {
    margin-right: 15px;
  }
}

.isRunning {
  color: $blue;
  border: 1px solid $blue;
  // background: linear-gradient(90deg, $blue, white);
}

.isPending {
  color: $gray;
  border: 1px solid $gray;
}

.isRejected {
  color: $red;
  border: 1px solid $red;
  // background: linear-gradient(90deg, $red, white);
}

.isSuccess {
  color: $green;
  border: 1px solid $green;
  // background: linear-gradient(90deg, $green, white);
}
