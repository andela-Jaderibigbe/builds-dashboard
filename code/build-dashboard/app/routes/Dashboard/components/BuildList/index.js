import React, { Component } from 'react';
import PropTypes from 'prop-types';
import TableBody from 'components/TableBody';
import BuildItem from '../BuildItem';

import styles from './BuildList.scss';

class BuildList extends Component {
  constructor(props) {
    super(props);

    this.state = {
      currentRow: null,
    };
  }

  onSelectRow = rowId => {
    let { currentRow } = this.state;
    currentRow = currentRow === rowId ? null : rowId;
    this.setState({ currentRow });
  };

  renderItems() {
    const { data } = this.props;
    const { currentRow } = this.state;

    return data.map(item => (
      <BuildItem
        key={item._id}
        {...item}
        isSelected={item._id === currentRow}
        onSelectItem={this.onSelectRow}
      />
    ));
  }

  render() {
    return <TableBody className={styles.tableBody}>{this.renderItems()}</TableBody>;
  }
}

BuildList.propTypes = {
  data: PropTypes.array,
};

export default BuildList;
