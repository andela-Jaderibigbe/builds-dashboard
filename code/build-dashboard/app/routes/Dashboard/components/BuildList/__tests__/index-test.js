import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import { expect as chaiExpect } from 'chai';
import data from 'data.json';

import BuildItem from 'routes/Dashboard/components/BuildItem';
import BuildList from '../';

describe('<BuildList />', () => {
  it('should render correctly', () => {
    const tree = renderer.create(<BuildList data={data} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should have correct number of rows', () => {
    const wrapper = shallow(<BuildList data={data} />);
    chaiExpect(wrapper).to.exist;
    chaiExpect(wrapper.find(BuildItem)).to.have.length(6);
  });
});
