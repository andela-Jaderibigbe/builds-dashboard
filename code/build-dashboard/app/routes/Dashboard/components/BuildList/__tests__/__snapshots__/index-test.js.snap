// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BuildList /> should render correctly 1`] = `
<div
  className="table tableBody"
>
  <div
    className="tableRow tableRow isPending"
  >
    <div
      className="cell changeListCol"
    >
      <span
        className="buildIcon"
      >
        <i
          className="fa fa-hourglass-start"
        />
        <span
          className="buildName"
          onClick={[Function]}
        >
          Tenrox-R1_1235
        </span>
      </span>
    </div>
    <div
      className="cell ownerCol"
    >
      
    </div>
    <div
      className="cell timeCol"
    />
    <div
      className="cell stateCol"
    >
      Pending
    </div>
    <div
      className="cell statusCol"
    >
      <div
        className="progressBar"
      >
        <span
          className="isDefault"
          style={
            Object {
              "opacity": 1,
              "width": "0%",
            }
          }
        />
      </div>
    </div>
    <div
      className="cell statusCol"
    >
      <div
        className="progressBar"
      >
        <span
          className="isDefault"
          style={
            Object {
              "opacity": 1,
              "width": "0%",
            }
          }
        />
      </div>
    </div>
    <div
      className="cell statusCol"
    >
      <div
        className="progressBar"
      >
        <span
          className="isDefault"
          style={
            Object {
              "opacity": 1,
              "width": "0%",
            }
          }
        />
      </div>
    </div>
    <div
      className="cell statusCol"
    >
      <div
        className="progressBar"
      >
        <span
          className="isDefault"
          style={
            Object {
              "opacity": 1,
              "width": "0%",
            }
          }
        />
      </div>
    </div>
  </div>
  <div
    className="tableRow tableRow isRunning"
  >
    <div
      className="cell changeListCol"
    >
      <span
        className="buildIcon"
      >
        <i
          className="fa fa-hourglass-half"
        />
        <span
          className="buildName"
          onClick={[Function]}
        >
          432462
        </span>
      </span>
    </div>
    <div
      className="cell ownerCol"
    >
      jack
    </div>
    <div
      className="cell timeCol"
    >
      10/17/2017 04:48 pm
    </div>
    <div
      className="cell stateCol"
    >
      Running
    </div>
    <div
      className="cell statusCol"
    >
      <div
        className="progressBar"
      >
        <span
          className="isRunning"
          style={
            Object {
              "opacity": 1,
              "width": "100%",
            }
          }
        />
      </div>
    </div>
    <div
      className="cell statusCol"
    >
      <div
        className="progressBar"
      >
        <span
          className="isRunning"
          style={
            Object {
              "opacity": 1,
              "width": "20%",
            }
          }
        />
      </div>
    </div>
    <div
      className="cell statusCol"
    >
      <div
        className="progressBar"
      >
        <span
          className="isRunning"
          style={
            Object {
              "opacity": 1,
              "width": "0%",
            }
          }
        />
      </div>
    </div>
    <div
      className="cell statusCol"
    >
      <div
        className="progressBar"
      >
        <span
          className="isRunning"
          style={
            Object {
              "opacity": 1,
              "width": "0%",
            }
          }
        />
      </div>
    </div>
  </div>
  <div
    className="tableRow tableRow isRejected"
  >
    <div
      className="cell changeListCol"
    >
      <span
        className="buildIcon"
      >
        <i
          className="fa fa-hourglass-end"
        />
        <span
          className="buildName"
          onClick={[Function]}
        >
          432461
        </span>
      </span>
    </div>
    <div
      className="cell ownerCol"
    >
      samy
    </div>
    <div
      className="cell timeCol"
    >
      10/13/2017 02:48 pm
    </div>
    <div
      className="cell stateCol"
    >
      Rejected
    </div>
    <div
      className="cell statusCol"
    >
      <div
        className="progressBar"
      >
        <span
          className="isError"
          style={
            Object {
              "opacity": 1,
              "width": "100%",
            }
          }
        />
      </div>
    </div>
    <div
      className="cell statusCol"
    >
      <div
        className="progressBar"
      >
        <span
          className="isSuccess"
          style={
            Object {
              "opacity": 1,
              "width": "100%",
            }
          }
        />
      </div>
    </div>
    <div
      className="cell statusCol"
    >
      <div
        className="progressBar"
      >
        <span
          className="isSuccess"
          style={
            Object {
              "opacity": 1,
              "width": "100%",
            }
          }
        />
      </div>
    </div>
    <div
      className="cell statusCol"
    >
      <div
        className="progressBar"
      >
        <span
          className="isSuccess"
          style={
            Object {
              "opacity": 1,
              "width": "100%",
            }
          }
        />
      </div>
    </div>
  </div>
  <div
    className="tableRow tableRow isSuccess"
  >
    <div
      className="cell changeListCol"
    >
      <span
        className="buildIcon"
      >
        <i
          className="fa fa-hourglass-end"
        />
        <span
          className="buildName"
          onClick={[Function]}
        >
          Tenrox-R1_1234
        </span>
      </span>
    </div>
    <div
      className="cell ownerCol"
    >
      
    </div>
    <div
      className="cell timeCol"
    >
      10/11/2017 12:48 pm
    </div>
    <div
      className="cell stateCol"
    >
      Complete
    </div>
    <div
      className="cell statusCol"
    >
      <div
        className="progressBar"
      >
        <span
          className="isSuccess"
          style={
            Object {
              "opacity": 1,
              "width": "100%",
            }
          }
        />
      </div>
    </div>
    <div
      className="cell statusCol"
    >
      <div
        className="progressBar"
      >
        <span
          className="isSuccess"
          style={
            Object {
              "opacity": 1,
              "width": "100%",
            }
          }
        />
      </div>
    </div>
    <div
      className="cell statusCol"
    >
      <div
        className="progressBar"
      >
        <span
          className="isSuccess"
          style={
            Object {
              "opacity": 1,
              "width": "100%",
            }
          }
        />
      </div>
    </div>
    <div
      className="cell statusCol"
    >
      <div
        className="progressBar"
      >
        <span
          className="isSuccess"
          style={
            Object {
              "opacity": 1,
              "width": "100%",
            }
          }
        />
      </div>
    </div>
  </div>
  <div
    className="tableRow tableRow isRejected"
  >
    <div
      className="cell changeListCol"
    >
      <span
        className="buildIcon"
      >
        <i
          className="fa fa-hourglass-end"
        />
        <span
          className="buildName"
          onClick={[Function]}
        >
          432460
        </span>
      </span>
    </div>
    <div
      className="cell ownerCol"
    >
      samy
    </div>
    <div
      className="cell timeCol"
    >
      07/11/2017 04:48 pm
    </div>
    <div
      className="cell stateCol"
    >
      Rejected
    </div>
    <div
      className="cell statusCol"
    >
      <div
        className="progressBar"
      >
        <span
          className="isError"
          style={
            Object {
              "opacity": 1,
              "width": "100%",
            }
          }
        />
      </div>
    </div>
    <div
      className="cell statusCol"
    >
      <div
        className="progressBar"
      >
        <span
          className="isDefault"
          style={
            Object {
              "opacity": 1,
              "width": "0%",
            }
          }
        />
      </div>
    </div>
    <div
      className="cell statusCol"
    >
      <div
        className="progressBar"
      >
        <span
          className="isDefault"
          style={
            Object {
              "opacity": 1,
              "width": "0%",
            }
          }
        />
      </div>
    </div>
    <div
      className="cell statusCol"
    >
      <div
        className="progressBar"
      >
        <span
          className="isDefault"
          style={
            Object {
              "opacity": 1,
              "width": "0%",
            }
          }
        />
      </div>
    </div>
  </div>
  <div
    className="tableRow tableRow isSuccess"
  >
    <div
      className="cell changeListCol"
    >
      <span
        className="buildIcon"
      >
        <i
          className="fa fa-hourglass-end"
        />
        <span
          className="buildName"
          onClick={[Function]}
        >
          432459
        </span>
      </span>
    </div>
    <div
      className="cell ownerCol"
    >
      samy
    </div>
    <div
      className="cell timeCol"
    >
      08/13/2017 01:48 pm
    </div>
    <div
      className="cell stateCol"
    >
      Accepted
    </div>
    <div
      className="cell statusCol"
    >
      <div
        className="progressBar"
      >
        <span
          className="isSuccess"
          style={
            Object {
              "opacity": 1,
              "width": "100%",
            }
          }
        />
      </div>
    </div>
    <div
      className="cell statusCol"
    >
      <div
        className="progressBar"
      >
        <span
          className="isSuccess"
          style={
            Object {
              "opacity": 1,
              "width": "100%",
            }
          }
        />
      </div>
    </div>
    <div
      className="cell statusCol"
    >
      <div
        className="progressBar"
      >
        <span
          className="isSuccess"
          style={
            Object {
              "opacity": 1,
              "width": "100%",
            }
          }
        />
      </div>
    </div>
    <div
      className="cell statusCol"
    >
      <div
        className="progressBar"
      >
        <span
          className="isSuccess"
          style={
            Object {
              "opacity": 1,
              "width": "100%",
            }
          }
        />
      </div>
    </div>
  </div>
</div>
`;
